### PR TITLE
add new exchange mexc

### PR DESCRIPTION
**Pull request by MXCDev**
**Before submitting this PR, please make sure**:

* [x] Your code builds clean without any errors or warnings
* [x] You are using approved title ("feat/", "fix/", "docs/", "refactor/")

**A description of the changes proposed in the pull request**:

**Tests performed by the developer**:

**Tips for QA testing**:

┆Issue is synchronized with this [Clickup task](https://app.clickup.com/t/1tjhju7) by [Unito](https://www.unito.io)

### CODE REVIEW

**Comment by dennisocana** [comment]
@MXCDev 

image

**Review by dennisocana — CHANGES_REQUESTED** [review]
@MXCDev please add `mexc_api_key` and `mexc_secret_key` in [conf_global_TEMPLATE.yml](https://github.com/mxcdevelop/hummingbot/blob/feat/add_new_exchange_mexc_global/hummingbot/templates/conf_global_TEMPLATE.yml)

image

**Comment by MXCDev** [comment]
@dennisocana  hi. we add unit test. pls approval auto test

**Comment by rxlxrxsx** [comment]
- Inconsistent order creation due to delay (around 1s) in balance update. See clip attached for reference.

https://user-images.githubusercontent.com/64377055/141056244-92e6de13-8c95-46f4-90b1-34ea32514e87.mp4

Logs and config: [mexc_delay.zip](https://github.com/CoinAlpha/hummingbot/files/7510144/mexc_delay.zip)

**Comment by rxlxrxsx** [comment]
> Inconsistent order creation due to delay (around 1s) in balance update. See clip attached for reference.

This is confirmed fixed.

- Another issue found - When restarting the strategy with hanging orders enabled, the cancelled orders after stopping the strategy are recognized as hanging orders that are cancelled after restart.

Screen Shot 2021-11-14 at 9 33 47 AM

Screen Shot 2021-11-14 at 9 34 22 AM

Logs and config: [mexc.zip](https://github.com/CoinAlpha/hummingbot/files/7533349/mexc.zip)

**Comment by MXCDev** [comment]
@rxlxrxsx  fixed.

**Review by rxlxrxsx — APPROVED** [review]
Tested and confirmed working by QA. No other issues found during testing.

**Review by dennisocana — APPROVED** [review]
Awaiting review and approval from engineering team.

File: hummingbot/connector/exchange/mexc/mexc_api_user_stream_data_source.py [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+        return cls._mexcausds_logger
+
+    def __init__(self, mexc_auth: MexcAuth, trading_pairs: Optional[List[str]] = []):
+        self._current_listen_key = None
+        self._current_endpoint = None
+        self._listen_for_user_stram_task = None
+        self._last_recv_time: float = 0
+        self._auth: MexcAuth = mexc_auth
+        self._trading_pairs = trading_pairs
+        super().__init__()
```
Comment (CrimsonJacket) [inline hummingbot/connector/exchange/mexc/mexc_api_user_stream_data_source.py:48]: As much as possible, lets use the same `_shared_client: aiohttp.ClientSession` from `MexcExchange`.

File: hummingbot/templates/conf_global_TEMPLATE.yml [hunk 2]
```diff
@@ -127,6 +127,9 @@ bybit_perpetual_secret_key: null
 bybit_perpetual_testnet_api_key: null
 bybit_perpetual_testnet_secret_key: null
 
+mexc_api_key: null
+mexc_secret_key: null
+
```
Comment (CrimsonJacket) [inline hummingbot/templates/conf_global_TEMPLATE.yml:132]: Do update the `template_version` to `29` at the top of the TEMPLATE file.

**Review by CrimsonJacket — CHANGES_REQUESTED** [review]
I noticed that some components are implemented in Cython, we're aiming for a Python-only connector architecture so as much as possible please do use Python. 

Also I noticed that you've decided to use fixtures along with `MockWebServer` for testing, I would request the usage of `aioresponses` and our own `NetworkMockingAssistant` class for unit testing. Refer to Ndax connector and its unit tests if needed.

Left some additional comments wrt usage of the same `aiohttp.ClientSession` across the Exchange, APIOrderBookDataSource and UserStreamDataSource components. Again refer to Ndax connector.

Minor issue with updating the `template_version` in `conf_global_TEMPLATE.yml `

**Comment by CrimsonJacket** [comment]
@MXCDev After some discussion with the team, I'd like to clarify that a full Python implementation is optional.

**Review by CrimsonJacket — APPROVED** [review]
Changes LGTM.

I will also like to take this chance to include the changes/improvements to expect in v0.47
1. Python-only implementation
2. Improved unit testing
3. Improved handling/utilization of `aiohttp` websocket connections
4. Better management of `aiohttp.ClientSession` across all connector components
5. Integration of `AsyncThrottler` with MEXC spot exchange connector

**Review by PtrckM — APPROVED** [review]
Tested and confirmed working by QA. No issue found during retesting of the changes applied.

**Pull request merged**